### ISSUE TITLE
Document Interface bugs tests report 

### ISSUE THREAD

**Issue by Francesca-Bit** [terminusdb/terminusdb#383]
**Describe the bug**
Problem in Insert/Update the database using the document interface

### First test
**To Reproduce**
1. Add the following json class with the property @documentation in the database schema **--ADD DOCUMENT POST CALL--**
`[{"@type":"Class","@id":"Doc01","@key":{"@type":"Random"},"num01":"xsd:byte","@documentation":{"@properties":{"num01":"hello"}}}]`

2. Call The GET method to view the database schema 
```
[{ "@base":"terminusdb:///data/",
  "@schema":"terminusdb:///schema#",
  "@type":"@context"
},
{
  "@id":"Doc01",
  "@key": {"@type":"Random"},
  "@type":"Class",
  "num01":"xsd:byte"
}
]
```
PROBLEM : the @documentation does not appear

### Second Test
**To Reproduce**
1. Update the Doc01 using the following json class  **--UPDATE DOCUMENT PUT CALL--**
[{"@id":"Doc01","@key":{"@type":"Random"},"@type":"Class","num01":"xsd:byte","@documentation":{"@comment":"Doc description","@properties":{"num01":"num01 description"}}}]

2. Call The GET method to view the database schema 
```
[
{
  "@base":"terminusdb:///data/",
  "@schema":"terminusdb:///schema#",
  "@type":"@context"
}Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Credentials: true
Access-Control-Max-Age: 1728000
Access-Control-Allow-Headers: Authorization, Authorization-Remote, Accept, Accept-Encoding, Accept-Language, Host, Origin, Referer, Content-Type, Content-Length, Content-Range, Content-Disposition, Content-Description
Access-Control-Allow-Origin: http://localhost:3030
Status: 500
Content-type: application/json

{
  "api:message":"Error: duplicate_key('terminusdb:///schema#num01')\ncontext(system:dict_pairs/3,_22252)\n",
  "api:status":"api:server_error"
}
```
PROBLEM the database schema is broken

**Comment by spl** [comment]
For this first test (using [xh](https://github.com/ducaale/xh)):

```
$ xh 'http://localhost:6363/api/document/admin/asdf?author=test&message=test&graph_type=schema' << EOF
> {"@type":"Class","@id":"Doc01","@key":{"@type":"Random"},"num01":"xsd:byte","@documentation":{"@properties":{"num01":"hello"}}}
> EOF
HTTP/1.1 200 OK
connection: keep-alive
content-length: 11
content-type: application/json; charset=UTF-8
date: Tue, 10 Aug 2021 19:29:59 GMT

[
    "Doc01"
]

$ xh 'http://localhost:6363/api/document/admin/asdf?author=test&message=test&graph_type=schema'
HTTP/1.1 200 OK
connection: keep-alive
content-length: 244
content-type: application/json; stream=true; charset=UTF-8
date: Tue, 10 Aug 2021 19:30:29 GMT

{
  "@base":"http://blah/base",
  "@schema":"http://blah/schema",
  "@type":"@context",
  "doc":"terminusdb:///data/",
  "scm":"terminusdb:///schema#"
}
{
  "@id":"Doc01",
  "@key": {"@type":"Random"},
  "@type":"Class",
  "num01":"xsd:byte"
}
````

**Comment by spl** [comment]
Is it somewhere here?

```prolog
id_schema_json(DB, Prefixes, Id, JSON) :-
    database_schema(DB,Schema),
    (   ground(Id)
    ->  prefix_expand_schema(Id, Prefixes, Id_Ex)
    ;   Id = Id_Ex
    ),

    xrdf(Schema, Id_Ex, rdf:type, Class),

    findall(
        K-V,
        (   distinct([P],xrdf(Schema,Id_Ex,P,O)),
            schema_subject_predicate_object_key_value(DB,Prefixes,Id_Ex,P,O,K,V)
        ),   
        Data),
    !,   
    compress_schema_uri(Id_Ex, Prefixes, Id_Compressed),
    compress_schema_uri(Class, Prefixes, Class_Compressed),
    (   atom_concat('sys:',Small_Class, Class_Compressed)
    ->  true 
    ;   Small_Class = Class_Compressed),
    dict_create(JSON,json,['@id'-Id_Compressed,
                           '@type'-Small_Class
                           |Data]).
```

**Comment by spl** [comment]
First one is fixed in 4be8175cd46e31a330f723d3e7bb936c11a55a41.

**Comment by spl** [comment]
@Francesca-Bit To elaborate, we now throw an error if no `@comment` is included in the `@documentation`.

**Comment by spl** [comment]
To reproduce the second issue:

```
$ ./terminusdb db create asdf -l "asdf" -c "asdf" --schema true --prefixes '{"@base":"http://blah/base","@schema":"http://blah/schema"}'

$ xh 'http://localhost:6363/api/document/admin/asdf?author=test&message=test&graph_type=schema' << EOF
> {"@type":"Class","@id":"Doc01","@key":{"@type":"Random"},"num01":"xsd:byte","@documentation":{"@properties":{"num01":"hello"},"@comment":"this is a comment"}}
> EOF
HTTP/1.1 200 OK
connection: keep-alive
content-length: 11
content-type: application/json; charset=UTF-8
date: Wed, 11 Aug 2021 08:28:39 GMT

[
    "Doc01"
]

$ xh PUT 'http://localhost:6363/api/document/admin/asdf?author=test&message=test&graph_type=schema' << EOF
> {"@id":"Doc01","@key":{"@type":"Random"},"@type":"Class","num01":"xsd:byte","@documentation":{"@comment":"Doc description","@properties":{"num01":"num01 description"}}}
> EOF

$ xh 'http://localhost:6363/api/document/admin/asdf?author=test&message=test&graph_type=schema' 
HTTP/1.1 200 OK
connection: keep-alive
content-length: 340
content-type: application/json; stream=true; charset=UTF-8
date: Wed, 11 Aug 2021 08:30:38 GMT

{
  "@base":"http://blah/base",
  "@schema":"http://blah/schema",
  "@type":"@context",
  "doc":"terminusdb:///data/",
  "scm":"terminusdb:///schema#"
}
Status: 500
Content-type: application/json

{
  "api:message":"Error: duplicate_key('http://blah/schemanum01')\ncontext(system:dict_pairs/3,_11496)\n",
  "api:status":"api:server_error"
}
```

**Comment by spl** [comment]
1. This appears to be fixed as of f8833762796b7ce84f4cc13e5e244c0941f24866.
2. Since this error is due to a bug, we plan to not have any future server errors in the middle of a JSON stream, because we will not have any bugs!

**Comment by spl** [comment]
This is what the second issue response now looks like:
```
$ xh 'http://localhost:6363/api/document/admin/testdb?author=author&message=message&graph_type=schema'
HTTP/1.1 200 OK
connection: keep-alive
content-length: 314
content-type: application/json; stream=true; charset=UTF-8
date: Wed, 11 Aug 2021 12:34:38 GMT

{
  "@base":"http://base/",
  "@schema":"http://schema/",
  "@type":"@context",
  "doc":"terminusdb:///data/",
  "scm":"terminusdb:///schema#"
}
{
  "@documentation": {"@comment":"comment", "@properties": {"num01":"hello"}},
  "@id":"Doc01",
  "@key": {"@type":"Random"},
  "@type":"Class",
  "num01":"xsd:byte"
}
```

**Comment by Francesca-Bit** [comment]
using Terminusdb branch dev (server)
 
If I try to update (PUT call) a schema graph class with a @documentation field 

Here the step to reproduce it :
1) create a database
2) add a new document 
`{
  "@type": "Class",
  "@id": "Doc001",
  "@key": {
    "@type": "Random"
  },
  "@documentation": {
    "@comment": "comment 01"
  }
}`
3) update the new document (change the @documentation value)
`{
  "@type": "Class",
  "@id": "Doc001",
  "@key": {
    "@type": "Random"
  },
  "@documentation": {
    "@comment": "comment 01 02"
  }
}`

4) the update call is ok (I got back the [Doc001])
5) The error happen when you try to get the schema back (GET CALL)

http://127.0.0.1:6363/api/document/admin/mytest/local/branch/main?as_list=true&graph_type=schema

if you try to get the database schema, you get the following error instead of the json schema
`[
{
  "@base":"terminusdb:///data/",
  "@schema":"terminusdb:///schema#",
  "@type":"@context"
}Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Credentials: true
Access-Control-Max-Age: 1728000
Access-Control-Allow-Headers: Authorization, Authorization-Remote, Accept, Accept-Encoding, Accept-Language, Host, Origin, Referer, Content-Type, Content-Length, Content-Range, Content-Disposition, Content-Description
Access-Control-Allow-Origin: http://localhost:3030
Status: 500
Content-type: application/json

{
  "api:message":"Error: duplicate_key('@documentation')\n  [36] dict_create(_6448,json,['@id'-'Doc002',...|...])\n  [33] api_document:api_generate_documents_(schema,transaction_object{commit_info:commit_info{},descriptor:branch_descriptor{branch_name:\"main\",repository_descriptor: ...},inference_objects:[],instance_objects:[...],parent:transaction_object{descriptor: ...,inference_objects:[],instance_objects: ...,parent: ...,schema_objects: ...},schema_objects:[...]},true,true,0,unlimited,_6518) at /app/terminusdb/src/core/api/api_document.pl:77\n  [31] '$apply':forall('<garbage_collected>',routes:json_write_with_header(...,_6648,...,true,[])) at /usr/lib/swipl/boot/apply.pl:52\n  [30] '<meta-call>'('<garbage_collected>') <foreign>\n  [29] catch(routes:(...,...),error(duplicate_key('@documentation'),context(...,_6750)),routes:do_or_die(...,...)) at /usr/lib/swipl/boot/init.pl:532\n  [28] catch_with_backtrace('<garbage_collected>','<garbage_collected>','<garbage_collected>') at /usr/lib/swipl/boot/init.pl:582\n\nNote: some frames are missing due to last-call optimization.\nRe-run your program in debug mode (:- debug.) to get more detail.\n\n",
  "api:status":"api:server_error"
}`

**Comment by GavinMendelGleason** [comment]
Fixed in 2905d0e49c124ac67b69e6f7a2db9cd01d094cd2
